feat(mobile): lazy loading of assets

fyfrey opened the pull request:
reworks how asset metadata is loaded in the app

allows to display huge asset collections on timeline and in albums

- [x] all screens now use the asset grid
- [x] no longer loads all assets to render the list. instead batches are fetched from the local database as needed
- [x] simplified state management of assets: all used providers are directly updated when the Isar database changes
- [x] removes 500+ LOC :-)
- [x] adds a new asset grouping mode "automatic": if there are few assets in a month, groups by month. If there are many, groups by day within that month. If there are too many assets on a single day, split into time ranges
- [x] album header now shrinks in size when asset grid is scrolled
- [x] favorite provider is no longer needed to render assets: assets have `isFavorite`
- [x] asset detail viewer watches asset for changes (e.g. sync running in background and isFavorite is changed OR it is merged with remote info etc.)

fyfrey committed:
feat(mobile): lazy loading of assets

fyfrey committed:
remove loading all assets upfront

fyfrey committed:
await batch-edit actions, disable controls while action is being processed

fyfrey committed:
display search results

fyfrey committed:
use lazy-loaded grid in album view

fyfrey committed:
re-enable dynamic grid layout

fyfrey committed:
delete unused file

fyfrey committed:
show albums without date groups

fyfrey committed:
fix last month not merged

fyfrey committed:
use db watcher for albums

fyfrey committed:
use pull-to-refresh to update timeline view

fyfrey committed:
simplify favorite provider

fyfrey committed:
fix unit tests

fyfrey committed:
tweak auto asset grouping

fyfrey committed:
invalidate settings provider

fyfrey committed:
format time range on same day

fyfrey committed:
fix rebase issue

matthinc commented:
![Screenshot 2023-05-09 at 19 10 06](https://github.com/immich-app/immich/assets/6898797/eb52fad8-5ccc-47be-bfcc-3d7bbec4026a)

Little display bug while drag-scrolling

matthinc commented on mobile/lib/modules/favorite/providers/favorite_provider.dart:
```diff
@@ -1,68 +1,77 @@
-      .allAssets
-      .where((element) => favorites.contains(element.id))
-      .toList();
+final favoriteAssetsProvider = StreamProvider<RenderList>((ref) async* {
+  final query = ref
+      .watch(dbProvider)
+      .assets
+      .filter()
+      .ownerIdEqualTo(Store.get(StoreKey.currentUser).isarId)
+      .isFavoriteEqualTo(true)
+      .sortByFileCreatedAt();
+  final settings = ref.watch(appSettingsServiceProvider);
+  yield await RenderList.fromQuery(
+    query,
+    GroupAssetsBy.values[settings.getSetting(AppSettingsEnum.groupAssetsBy)],
+  );
+  await for (final _ in query.watchLazy()) {
+    yield await RenderList.fromQuery(
+      query,
+      GroupAssetsBy.values[settings.getSetting(AppSettingsEnum.groupAssetsBy)],
+    );
+  }
```
Cool!

matthinc commented on mobile/lib/modules/home/ui/asset_grid/asset_grid_data_structure.dart:
```diff
@@ -2,212 +2,282 @@ import 'dart:math';
+
+  bool get isEmpty => totalAssets == 0;
+
+  List<Asset> loadAssets(int offset, int count) {
+    assert(offset >= 0);
+    assert(count > 0);
+    assert(offset + count <= totalAssets);
+    if (allAssets != null) {
+      return allAssets!.slice(offset, offset + count);
+    } else if (query != null) {
+      if (offset < _bufOffset || offset + count > _bufOffset + _buf.length) {
+        final bool forward = _bufOffset < offset;
+        const batchSize = 128;
+        const oppositeSize = 32;
+        final len = max(batchSize, count + oppositeSize);
+        final start = max(
+          0,
+          forward
+              ? offset - oppositeSize
+              : (len > batchSize ? offset : offset + count - len),
+        );
+        _buf = query!.offset(start).limit(len).findAllSync();
+        _bufOffset = start;
+      }
+      assert(_bufOffset <= offset);
+      assert(_bufOffset + _buf.length >= offset + count);
+      final arr = _buf.slice(offset - _bufOffset, offset - _bufOffset + count);
+      return arr;
     }
+    throw Exception("RenderList has neither assets nor query");
+  }
```
Can you add some comments explaining your algorithm, please? This code is (at least for me) very complex and hard to understand.

fyfrey replied:
sure, added comments

matthinc commented:
I really like the approach of lazy-loading assets only when needed. We probably have to come up with another drag-scrolling mechanism since the current one relies on data from all assets to build its layout properly. Do you have any idea in mind for alternatives?

fyfrey commented:
> ![Screenshot 2023-05-09 at 19 10 06](https://user-images.githubusercontent.com/6898797/237169785-eb52fad8-5ccc-47be-bfcc-3d7bbec4026a.png)
> 
> Little display bug while drag-scrolling

What's the bug? showing simple bars instead of laying out all assets is a necessary simplification as we don't know the assets while scrolling :-)

matthinc commented:
> > ![Screenshot 2023-05-09 at 19 10 06](https://user-images.githubusercontent.com/6898797/237169785-eb52fad8-5ccc-47be-bfcc-3d7bbec4026a.png)
> > Little display bug while drag-scrolling
> 
> What's the bug? showing simple bars instead of laying out all assets is a simplification as we don't know the assets while scrolling :-) we could of course always make square boxes according to the number of assets per row setting.. I'll look into that. If there's another issue please notify me

Ok, those boxes looked buggy to me as I knew the previous behavior. I don't think there is any point in showing them at all if they don't represent anything. What about just leaving the screen black then?

fyfrey commented:
> Ok, those boxes looked buggy to me as I knew the previous behavior. I don't think there is any point in showing them at all if they don't represent anything. What about just leaving the screen black then?

We need something to fill the exact height so the scrolling behaves correctly. Showing nothing (black screen) will probably be weird while scrolling a long time without headlines from days/months. I'll rework this to be similar to the previous behavior.

fyfrey committed:
review feedback

fyfrey committed:
try to fix mobile unit test execution

fyfrey committed:
try to fix mobile unit tests again

fyfrey commented on mobile/lib/modules/album/views/album_viewer_page.dart:
```diff
@@ -241,75 +200,71 @@ class AlbumViewerPage extends HookConsumerWidget {
+            },
+            child: Column(
+              children: [
+                // TODO header takes up space while scrolling through the assets
```
@martyfuhry can you take a look at this issue if you have some time? thanks!

martyfuhry committed:
Sliver app bar for album page (#2431)

* Sliver app bar for album page

* Adds in check for remote album

* Fixes bug where multiselect can be in a weird state

fyfrey committed:
manually scroll header in nested scroll view

martyfuhry commented on mobile/lib/modules/home/ui/asset_grid/immich_asset_grid_view.dart:
```diff
@@ -241,14 +333,33 @@ class ImmichAssetGridViewState extends State<ImmichAssetGridView> {
+    final end = values.lastOrNull;
+    if (start != null && end != null) {
+      if (start.index <= end.index) {
+        widget.visibleItemsListener!(start, end);
```
`widget.visibleItemsListener?.call(start, end)` is safer. Even though this is properly guarded at line 344, the optional `.?call` will be more maintainable in the future.

martyfuhry commented on mobile/lib/modules/home/ui/asset_grid/immich_asset_grid_view.dart:
```diff
@@ -241,14 +333,33 @@ class ImmichAssetGridViewState extends State<ImmichAssetGridView> {
+      if (start.index <= end.index) {
+        widget.visibleItemsListener!(start, end);
+      } else {
+        widget.visibleItemsListener!(end, start);
```
Same here.

fyfrey committed:
review feedback

fyfrey committed:
rework album page to best match old behavior

fyfrey committed:
fix asset grid album rendering

fyfrey committed:
remove code duplication

fyfrey committed:
provider code readability

fyfrey committed:
remove unneeded favorite probider

alextran1502 committed:
dev: styling date block

alextran1502 commented:
Hello Fynn, I've done my bit of testing, and below are a few things I found.

1. During the database indexing, and upserting process upon a new login, we don't have any indication. This might confuse people.
2. When switching users by logging out and logging in the newly logged-in user has the database indexing completed but doesn't have the timeline filled in, as well as the library page. The sharing page correctly displays the albums. In order to get the timeline to show the asset and the library page to show the albums I have to restart the app. This might be the problem with the stream listener?

Let me know if you need anything else

fyfrey committed:
fix showing loading indicator on home screen

fyfrey committed:
automatically dispose providers to fix user change issue

fyfrey committed:
optimize album header scrolling behavior

fyfrey commented:
> 1. During the database indexing, and upserting process upon a new login, we don't have any indication. This might confuse people.

Indeed. I've fixed this by showing the loading indicator instead if the grid would be empty 

> 2. When switching users by logging out and logging in the newly logged-in user has the database indexing completed but doesn't have the timeline filled in, as well as the library page. The sharing page correctly displays the albums. In order to get the timeline to show the asset and the library page to show the albums I have to restart the app. This might be the problem with the stream listener?

This was tricky. The providers are not normally reset while the app stays open... And the user id is thus cached in it and stays the same after changing logged in user. I've fixed this by make the providers auto dispose, so that they are destroyed when the view they're used on is destroyed.

alextran1502 commented on mobile/lib/shared/providers/asset.provider.dart:
```diff
@@ -220,80 +120,79 @@ class AssetNotifier extends StateNotifier<AssetsState> {
-    (a, b) => b.compareTo(a),
-  );
+final assetsProvider = StreamProvider.autoDispose<RenderList>((ref) async* {
+  final query = ref
+      .watch(dbProvider)
+      .assets
+      .filter()
+      .ownerIdEqualTo(Store.get(StoreKey.currentUser).isarId)
+      .isArchivedEqualTo(false)
+      .sortByFileCreatedAtDesc();
+  final settings = ref.watch(appSettingsServiceProvider);
+  final groupBy =
+      GroupAssetsBy.values[settings.getSetting(AppSettingsEnum.groupAssetsBy)];
+  yield await RenderList.fromQuery(query, groupBy);
+  await for (final _ in query.watchLazy()) {
+    yield await RenderList.fromQuery(query, groupBy);
+  }
```
I saw this implementation across the PR. Am I understanding correctly that on line 175 we return the result the first time this provider is watched/called? Then we setup the a listener to the subsequent changes on line 176-178 for any future changes?

fyfrey replied:
Yeah, that is correct!

alextran1502 committed:
Remove selection indicator in album asset selection view

alextran1502 commented:
I found a bug that when you click on favorite button for the first time, it showed your an error below

```
I/flutter (18024): [SEVERE] [2023-05-16 10:53:23.546425] 'package:immich_mobile/shared/models/asset.dart': Failed assertion: line 236 pos 12: 'canUpdate(a)': is not true.
```

Then, clicking on it again would work.

alextran1502 reviewed (approved):
Thank you for another high-quality PR. I only have a few more questions, and this should be good to merge!

fyfrey committed:
improve album error page

fyfrey commented on mobile/lib/modules/album/providers/shared_album.provider.dart:
```diff
@@ -75,21 +56,32 @@ class SharedAlbumNotifier extends StateNotifier<List<Album>> {
-  return a;
+  await for (final a in sharedAlbumService.watchAlbum(albumId)) {
+    if (a == null) {
+      throw Exception("Album with ID=$albumId does not exist anymore!");
```
No, it prints the error on the screen and the user can navigate back to the previous page.

fyfrey committed:
check canUpdate of asset

fyfrey commented:
> Thank you for another high-quality PR. I only have a few more questions, and this should be good to merge!

Added minor improvements for the two points you raised. This should be good to merge now.

alextran1502 commented:
@fyfrey I tested the changes you made, and it appears to cause some issues with rerendering/updating the page when an asset is archived/unarchived; Also, when clicking on the favorite button, the change isn't updated even though there is a message of `upserted x assets into the DB`, you will have to tap on the icon again for it to execute.

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into dev/lazy-loading
